[clang] Add source range to 'use of undeclared identifier' diagnostics

tbaederr opened the pull request.

tbaederr committed:
[clang] Add source range to 'use of undeclared identifier' diagnostics

tbaederr commented:
I knew there was something oddly familiar about this: https://reviews.llvm.org/D150191?id=521215 :man_facepalming:

cor3ntin commented:
Can you add a description? Maybe a changelog

tbaederr commented on clang/lib/Frontend/DiagnosticRenderer.cpp:
```diff
@@ -487,14 +487,17 @@ static bool checkRangeForMacroArgExpansion(CharSourceRange Range,
+  // llvm::errs() << __PRETTY_FUNCTION__ << '\n';
   assert(Loc.isMacroID() && "Must be a macro expansion!");
 
   SmallVector<CharSourceRange, 4> SpellingRanges;
-  mapDiagnosticRanges(Loc, Ranges, SpellingRanges);
+  // mapDiagnosticRanges(Loc, Ranges, SpellingRanges);
+  // llvm::errs() << "Spelling Ranges: " << SpellingRanges.size() << '\n';
 
   // Count all valid ranges.
-  unsigned ValidCount =
-      llvm::count_if(Ranges, [](const auto &R) { return R.isValid(); });
+  unsigned ValidCount = 0;
+      // llvm::count_if(Ranges, [](const auto &R) { return R.isValid(); });
+  // llvm::errs() << "ValidCount: " << ValidCount << '\n';
```
That was just a test commit, but didn't end up working out anyway. :(

tbaederr commented:
It's the same problem again.

For this code:
```c++

#define F(x) x + 1
#define G(x) F(x) + 2
#define ADD(x,y) G(x) + y
#define LEVEL4(x) ADD(p,x)
#define LEVEL3(x) LEVEL4(x)
#define LEVEL2(x) LEVEL3(x)
#define LEVEL1(x) LEVEL2(x)

int a = LEVEL1(b);
```

The previous output was:
```
./array.cpp:34:9: error: use of undeclared identifier 'p'
   34 | int a = LEVEL1(b);
      |         ^
./array.cpp:32:19: note: expanded from macro 'LEVEL1'
   32 | #define LEVEL1(x) LEVEL2(x)
      |                   ^
./array.cpp:31:19: note: expanded from macro 'LEVEL2'
   31 | #define LEVEL2(x) LEVEL3(x)
      |                   ^
./array.cpp:30:19: note: expanded from macro 'LEVEL3'
   30 | #define LEVEL3(x) LEVEL4(x)
      |                   ^
./array.cpp:29:23: note: expanded from macro 'LEVEL4'
   29 | #define LEVEL4(x) ADD(p,x)
      |                       ^
./array.cpp:34:16: error: use of undeclared identifier 'b'
   34 | int a = LEVEL1(b);
      |                ^
2 errors generated.
```

and we now print:
```
./array.cpp:34:9: error: use of undeclared identifier 'p'
   34 | int a = LEVEL1(b);
      |         ^~~~~~~~~
./array.cpp:32:19: note: expanded from macro 'LEVEL1'
   32 | #define LEVEL1(x) LEVEL2(x)
      |                   ^~~~~~~~~
./array.cpp:31:19: note: expanded from macro 'LEVEL2'
   31 | #define LEVEL2(x) LEVEL3(x)
      |                   ^~~~~~~~~
./array.cpp:30:19: note: expanded from macro 'LEVEL3'
   30 | #define LEVEL3(x) LEVEL4(x)
      |                   ^~~~~~~~~
note: (skipping 1 expansions in backtrace; use -fmacro-backtrace-limit=0 to see all)
./array.cpp:28:20: note: expanded from macro 'ADD'
   28 | #define ADD(x,y) G(x) + y
      |                    ^
./array.cpp:27:16: note: expanded from macro 'G'
   27 | #define G(x) F(x) + 2
      |                ^
./array.cpp:26:14: note: expanded from macro 'F'
   26 | #define F(x) x + 1
      |              ^
./array.cpp:34:16: error: use of undeclared identifier 'b'
   34 | int a = LEVEL1(b);
      |                ^
./array.cpp:32:26: note: expanded from macro 'LEVEL1'
   32 | #define LEVEL1(x) LEVEL2(x)
      |                          ^
./array.cpp:31:26: note: expanded from macro 'LEVEL2'
   31 | #define LEVEL2(x) LEVEL3(x)
      |                          ^
./array.cpp:30:26: note: expanded from macro 'LEVEL3'
   30 | #define LEVEL3(x) LEVEL4(x)
      |                          ^
./array.cpp:29:25: note: expanded from macro 'LEVEL4'
   29 | #define LEVEL4(x) ADD(p,x)
      |                         ^
./array.cpp:28:25: note: expanded from macro 'ADD'
   28 | #define ADD(x,y) G(x) + y
      |                         ^
2 errors generated.
```

but this is not a bug introduced by the changes, rather, the newly supplied (valid!) source ranges now get taken into account in `checkRangesForMacroArgExpansion()` in `Frontend/DiagnosticRenderer.cpp`.

AaronBallman commented:
> but this is not a bug introduced by the changes, rather, the newly supplied (valid!) source ranges now get taken into account in `checkRangesForMacroArgExpansion()` in `Frontend/DiagnosticRenderer.cpp`.

Yeah, I think we need to solve that before moving forward with the changes, which is why this keeps not being improved. (It always seems to come back to bite us when we decide that invalid source locations are an in-band indicator of anything.)

tbaederr committed:
Rewrite check* functions in DiagnosticRenderer.cpp

In rangeInsideSameMacroArgExpansion, check the given Ranges instead
of the SpellingRanges.

tbaederr commented:
This is the best version I could come up with. Other than that, the code as it is seems to do what is advertised, i.e. the new output is the desired output.

AaronBallman commented on clang/lib/Basic/SourceManager.cpp:
```diff
@@ -996,19 +996,22 @@ CharSourceRange SourceManager::getExpansionRange(SourceLocation Loc) const {
 bool SourceManager::isMacroArgExpansion(SourceLocation Loc,
                                         SourceLocation *StartLoc) const {
-  if (!Loc.isMacroID()) return false;
+  if (!Loc.isMacroID())
```
All the changes in this file can be reverted because they're just formatting changes, right?

AaronBallman commented on clang/lib/Frontend/DiagnosticRenderer.cpp:
```diff
@@ -454,62 +454,43 @@ void DiagnosticRenderer::emitSingleMacroExpansion(
-  // Count all valid ranges.
-  unsigned ValidCount =
-      llvm::count_if(Ranges, [](const auto &R) { return R.isValid(); });
+  {
```
Is the extra compound statement helpful?

AaronBallman reviewed (commented):
Generally speaking, this looks like it's heading in the right direction.

tbaederr commented:
Ping

AaronBallman reviewed (approved):
LGTM!
